### PR TITLE
Filter category

### PR DESCRIPTION
**Pull request by cailors**
-Select item rating...: return every items with any rating
-Select item type...: return every items with any type
-Clear Filters: return everything
Screen Shot 2022-07-26 at 8 45 23 AM

### CODE REVIEW

**Review by phreid — CHANGES_REQUESTED** [review]
Please fix merge conflicts (probably need to pull latest changes from dev branch).

**Comment by cailors** [comment]
> Please fix merge conflicts (probably need to pull latest changes from dev branch).

Done

File: src/server/routes/items.js [hunk 1]
```diff
@@ -150,5 +148,36 @@ router.get('/search/:searchText', async (req, res) => {
...
+ * Filtering items with selected inputs.
+ *
+ * @param filters: the JSON object that contains the filters
+ * @returns items
+ */
+router.get('/filter/:filters', async (req, res) => {
+  const rate = JSON.parse(req.params.filters).rating
+  const category = JSON.parse(req.params.filters).type
+  console.log(rate)
+  console.log(category)
```
Comment (phreid) [inline src/server/routes/items.js]: remove console.log statements

File: src/server/routes/items.js [hunk 2]
```diff
@@ -150,5 +148,36 @@ router.get('/search/:searchText', async (req, res) => {
...
+ * GET /items/filter/:filters
+ *
+ * Filtering items with selected inputs.
+ *
+ * @param filters: the JSON object that contains the filters
+ * @returns items
+ */
+router.get('/filter/:filters', async (req, res) => {
+  const rate = JSON.parse(req.params.filters).rating
+  const category = JSON.parse(req.params.filters).type
```
Comment (phreid) [inline src/server/routes/items.js:161]: if you're sending the filters as a json object (which is a fine thing to do i think), you should send them in the request body, not the url. then you'd get the object with `req.body`

File: src/redux/items/thunks.js [hunk 3]
```diff
@@ -67,3 +67,11 @@ export const applySearchNameAsync = createAsyncThunk(
     return response.data.result
   }
 )
+
+export const applyFiltersAsync = createAsyncThunk(
+  'items/filterItem',
+  async (filters) => {
+    const response = await axios.get(`${URL}/filter/${filters}`)
+    return response.data.result
```
Comment (phreid) [inline src/redux/items/thunks.js:75]: send the filters in the request body: 

`axios.get(${URL}/filter, filters)`

File: src/components/FiltersCollection.jsx [hunk 4]
```diff
@@ -1,12 +1,43 @@
...
+    setItemType(event.target.value)
+  }
+
+  function handleRating (event) {
+    setItemRate(event.target.value)
+  }
+
+  function handleApplyFilter () {
+    const jsonObj = JSON.stringify({ rating: itemRate, type: itemType })
+    console.log(jsonObj + '  line32')
```
Comment (phreid) [inline src/components/FiltersCollection.jsx]: remove console.log

File: src/styles/search.css [hunk 5]
```diff
@@ -51,15 +51,3 @@
 	margin-top: 35;
 	border-radius: 5;
 	background-color: #8B9CB6;
-}
```
Comment (phreid) [inline src/styles/search.css:54]: add this `}` back in

File: src/components/FiltersCollection.jsx [hunk 6]
```diff
@@ -1,12 +1,43 @@
...
+import { applySearchNameAsync, getAllItemsAsync, applyFiltersAsync } from '../redux/items/thunks'
 import { useDispatch } from 'react-redux'
 import { Button, ButtonToolbar, InputGroup, Form } from 'react-bootstrap'
+import { ITEM_TYPES, RATINGS } from '../constants'
 
 function FiltersCollection () {
   const dispatch = useDispatch()
   const [searchInput, setSearchInput] = useState('')
+  const [itemType, setItemType] = useState('Select item type...')
+  const [itemRate, setItemRate] = useState('Unrated')
```
Comment (phreid) [inline src/components/FiltersCollection.jsx]: You don't need to set the default message here. You can just add a default option like you have in the item type dropdown.

Comment (phreid) [inline src/components/FiltersCollection.jsx]: `itemRating`, `setItemRating`

File: src/server/routes/items.js [hunk 7]
```diff
@@ -150,5 +148,36 @@ router.get('/search/:searchText', async (req, res) => {
...
+    } else if (category === 'Select item type...') {
+      items = await Item.find({ rating: rate }).populate('owner')
+    } else {
+      items = await Item.find({ rating: rate, type: category }).populate('owner')
+    }
+    res.send({
+      result: items
+    })
+  }
+})
```
Comment (phreid) [inline src/server/routes/items.js]: this logic could be better. don't rely on the client sending the default values of 'Select item type...' etc. those values could change. 

instead check for null or undefined, and have the client send null or undefined in the filter object if they haven't picked a value

**Review by phreid — CHANGES_REQUESTED** [review]

File: src/components/FiltersCollection.jsx (hunk 6 above)
Comment (cailors) [reply src/components/FiltersCollection.jsx]: But the problem with that is after you click the second dropdown, the current dropdown goes back to the default option.

Comment (phreid) [reply src/components/FiltersCollection.jsx]: > But the problem with that is after you click the second dropdown, the current dropdown goes back to the default option.

Sorry not sure what you mean. If I change it to this everything works as expected: 

```
const [itemType, setItemType] = useState('')
const [itemRating, setItemRating] = useState('')
...
<Form.Group >
    <Form.Select value={itemType} onChange={handleCategory}>
      <option>Select item type...</option>
      {itemTypeDropdowns}
    </Form.Select>
  </Form.Group>
<Form.Group >
  <Form.Select value={itemRating} onChange={handleRating}>
    <option>Select a rating...</option>
    {ratingDropdowns}
  </Form.Select>
</Form.Group>
```

File: src/server/routes/items.js (hunk 2 above)
Comment (cailors) [reply src/server/routes/items.js:161]: I tried the to send inside the body, but I think it only accepts string. I will stay with my version for now

**Review by phreid — APPROVED** [review]
Made one small change to the default filter options.

File: src/components/FiltersCollection.jsx [hunk 8]
```diff
@@ -39,6 +88,24 @@ function FiltersCollection () {
...
+        </Form.Select>
+      </Form.Group>
+      <Form.Group >
+        <Form.Select value={itemRating} onChange={handleRating}>
+          <option>{DEFAULT_RATING_OPTION}</option>
+          {ratingDropdowns}
+        </Form.Select>
+      </Form.Group>
+      <br></br>
+      <a className = "searchBar">
```
Comment (imogenem) [inline src/components/FiltersCollection.jsx:105]: @cailors I'm getting a warning when running this, so I believe you just need to change this `a` tag to a `div`.  

The warning:
```
WARNING in src/components/FiltersCollection.jsx
  Line 105:7:  The href attribute is required for an anchor to be keyboard accessible. Provide a valid, navigable address as the href value. If you cannot provide an href, but still need the element to resemble a link, use a button and change it with appropriate styles

Comment (phreid) [reply src/components/FiltersCollection.jsx:105]: fixed

**Pull request merged**